### ISSUE TITLE
I have integrated api:paypal:4.8.1, but BrowserSwitchResult sometimes returns multiple times and there are errors

### ISSUE THREAD

**Issue by Yongyong-Zhou** [braintree/braintree_android#500]
![image](https://user-images.githubusercontent.com/86334406/150962256-d48d0ed5-24b5-4b33-90ec-88435d3cdb70.png)
![image](https://user-images.githubusercontent.com/86334406/150962288-cc9b6f05-8433-437b-b393-b31b5552de95.png)
![771643104184_ pic](https://user-images.githubusercontent.com/86334406/150962898-73dd1b73-ce59-4562-8c63-e4d83e84df35.jpg)

1.myTokenizePayPalAccountWithVaultMethod()、myTokenizePayPalAccountWithCheckoutMethod（）Is there any difference between these two?
2.I use myTokenizePayPalAccountWithCheckoutMethod() to pay, BrowserSwitchResult sometimes returns multiple times, and there are errors? Eg:
(1)The response contained inconsistent data.
(2)com.braintreepayments.api.usercanceledexception: user canceled paypal.

Please check my bug, I have been troubled for several days, thank you.. thank you

**Comment by sshropshire** [comment]
@Yongyong-Zhou thanks for your feedback. For the _user canceled paypal_ error response, that means the user returned to your app without completing the PayPal flow. For the first scenario, do you have steps to reproduce this issue? And how consistent is this error?

**Comment by indiascope** [comment]
Hi @sarahkoop  We are fetching the same issue.

We have Integrated **Lastest SDK (4.9.0**).  but in Android Device 9.0 and below first we are getting `onPayPalFailure()` callback and after this, we are getting `onPayPalSuccess()` callback. So we are getting `BrowserSwitchResult` multiple times. Above Android 10 and above it is working properly.

**Note**: Also, we have noticed strange behavior. In Real devices, the Android 9.0 activity lifecycle is different from than Android Emulator Android OS 9.0 and 8.0. So Please check in Real devices.

We have investigated this issue in detail  from Our end and found In `PayPalLifecycleObserver` class, when Activity or Fragment Lifecycle reach to `Resume` State,  then first `BrowserSwitchResult` is fetched and after this if request Code is **PayPal** then `PayPalClient`'s `onBrowserSwitchResult()` method is called. This method is internally calling another method `deliverBrowserSwitchResultToListener()` method. In this `onResult()` override method  of `PayPalBrowserSwitchResultCallback()`  SDK has register `onPayPalSuccess()` callback and `onPayPalFailure()` callback.

We found In **Android 9** and below Device `onResume()` method is calling before the `onNewIntent()` method. When `onResume() `method is called before `onNewIntent()` override method ` onPayPalFailure()`  callback is firing.  I think it is due to `shouldNotifyCancellation` value  . When  `BrowserSwitchRequest` object is created , then in  `shouldNotifyCancellation`  value is set to true. After `onNewIntent()` method, also again  `onResume()` method is calling in  **Android 9** and below. But at that time we can update the `intent` using `setIntent(intent)` in `onNewIntent()` override method  and if we update the intent then, we are getting  `onPayPalSuccess()` callback.

But In `Android 9` and below `onReusme()` is calling 2 times by Android OS, 1 is before `onNewIntent()` and 1 is after `onNewIntent()`   and due to this Behaviour, Our Users are confused and we are getting complaints from our users Why we are getting User Cancelled messages. In Actual a user doesn't cancel the payment flow and complete the flow and also purchase the item.

 We have noticed **Android 10** and above device `onResume()` method is always called after `onResume()`.  So Intent is updating and we are always getting `onPayPalSuccess()` callback.

If possible SDK Can internally call `getBrowserSwitchResult()` method after `onNewIntent()` method. So our Intent will be updated and  I think due to that  `BrowserSwitchResult`  will not call multiple times and we will don't get an `onPayPalFailure()` callback.

If it is can't be fixed by SDK or it will take a longer time then Please suggest how we can fixed it from our side.

```
// Paypal
implementation 'com.braintreepayments.api:paypal:4.9.0'
// Venmo
implementation 'com.braintreepayments.api:venmo:4.9.0'
```

**Android- 9 (Real Android Device)** 
```
Paypal onPause Called
Paypal onStop Called
Paypal onSaveInstanceState Called --Go to Paypal Page
Paypal onRestart Called --- Back from Paypal Page
Paypal onStart Called
Paypal onResume Called  --- onPayPalFailure() callback
Paypal onPause Called
Paypal onNewIntent Called
Paypal onResume Called  --- onPayPalSuccess() callback

```

**Android-10**
```
Paypal onPause Called
Paypal onStop Called
Paypal onSaveInstanceState Called --- Go to Paypal Page
Paypal onRestart Called --- Back from Paypal Page
Paypal onStart Called
Paypal onNewIntent Called
Paypal onResume Called --- onPayPalSuccess() callback

```

**Android-12**
```
Paypal onPause Called
Paypal onStop Called
Paypal onSaveInstanceState Called -- --- Go to Paypal Page Chrome Custom Tab
Paypal onRestart Called  --- Back from Paypal Page
Paypal onStart Called
Paypal onNewIntent Called
Paypal onResume Called -- --- onPayPalSuccess() callback
```

**Android-9.0(28) Emulator**

```
Paypal onPause Called
Paypal onStop Called
Paypal onSaveInstanceState Called-- --- Go to Paypal Page Chrome Custom Tab
Paypal onNewIntent Called --- Back from Paypal Page
Paypal onRestart Called
Paypal onStart Called
Paypal onResume Called -- --- onPayPalSuccess() callback
```

**Android-8(27) Emulator**
```
Paypal onPause Called
Paypal onSaveInstanceState Called
Paypal onStop Called-- --- Go to Paypal Page Chrome Custom Tab
Paypal onNewIntent Called --- Back from Paypal Page
Paypal onRestart Called
Paypal onStart Called
Paypal onResume Called -- --- onPayPalSuccess() callback
```

**Comment by sshropshire** [comment]
@indiascope thank you for providing details on this. We'll investigate and provide the necessary fixes.

**Comment by indiascope** [comment]
Thank you  @sshropshire for your reply. If you require any more information then please let me know. Also,  I have noticed strange behavior. In Real devices, the Android 9.0 activity lifecycle is different from than Android Emulator Android OS 9.0 and 8.0.

**Comment by sshropshire** [comment]
@indiascope while we're investigating a solution to this issue, is it possible in your app to ignore the `onPayPalFailure()` callback on devices using API 28 and below? We've seen similar issues in the past with Android 9 and the PayPal flow that are rooted in this Android bug that Google probably won't fix.

This bug makes it difficult for us to say for certain that the user has canceled the payment flow e.g. they could be multi-tasking and have intentions to return to the payment flow and complete the transaction.

**Comment by indiascope** [comment]
Sorry for asking.Any Updates on this issue @sshropshire @sarahkoop ?  We are getting so many complaints from Our Users. Most of the Users are using **Samsung** Device and Android Version is 9  or below 9.  Currently, I am thinking of 1 Hack. You can check and guide this will cover all scenarios or not. 

We have observed In **Samsung** Device Android Os 9 and below Android OS  is calling `OnResume()`  before `onNewIntent()` and also after `onNewIntent()`, `onResume()` is calling. Currently, I am thinking of putting a short delay in `public void onPayPalFailure(Exception error)` override method. I will put a delay in this `onPayPalFailure()` method only if  `onNewIntent()` override method is not called by Android OS. 

If `onNewIntent()` override method is called by Android OS then we will remove the delay and also we will make the boolean `isOnNewIntent` value to true. So the Failure callback will not be executed. 

After this,  in `onPayPalSuccess()` and `onPayPalFailure()` override method, due to `isOnNewIntent` value true, we will not put delay and we will directly show success, failure popup and also we will redirect a user to the proper screen. 

If the `onNewIntent()` method is not called by Android OS then maybe the user canceled the payment. Then in this scenario, after the fixed delay, the Error Message User Cancelled the Payment will be shown to the User.

What's Your suggestion. We are missing any scenario?  I personally don't like this hack. Any drawback in this hack? I think if this issue is fixed in the Library itself then it will be very better.

I have seen a similar solution is in discussion on a similar issue in  Auth0 SDK.
https://github.com/auth0/Auth0.Android/issues/530

**Comment by sshropshire** [comment]
@indiascope unfortunately a workaround may be the best solution here since Google probably won't fix. Thinking it through though, one of the solutions in the [Auth0](https://github.com/auth0/Auth0.Android/issues/530#issuecomment-981685713) thread may work.

Android activity lifecycle methods are called serially [according to SO](https://stackoverflow.com/a/30896358). I don't have an Android 9 Samsung device on hand, but I wonder if you could help us verify if this works?

```java
new Handler(Looper.getMainLooper()).post(() -> {
  BrowserSwitchResult browserSwitchResult = braintreeClient.deliverBrowserSwitchResult(this);
  if (browserSwitchResult != null) {
    payPalClient.onBrowserSwitchResult(activity, (payPalAccountNonce, error) -> {
      // continue transaction / handle error
    });
  }
});
```

Browser switch is designed to only deliver "once," so even if this `Runnable` ends up posting twice to the main thread, the second invocation will be a no-op.

**Comment by sshropshire** [comment]
Awesome! We may implement this fix internally and it will be available in future versions of the SDK. We may also provide some documentation for merchants using the legacy integration method.

The main `Looper` is associated with the Main UI thread and a serial message queue. When we call `post()`, the Runnable is queued up for execution on the next loop, which occurs after the current loop that is still processing `onResume` / `onNewIntent` calls.

The two invocations will run twice, one after the other. Our `BrowserSwitch` library will only "deliver" the result once. The second invocation of the Runnable results in a no-op for the SDK. It isn't ideal, but we can lean on this behavior to workaround Google's unpredictable lifecycle calls on Android 9. 

I've only scratched the surface, here's a deeper dive on [Android concurrency](https://blog.mindorks.com/android-core-looper-handler-and-handlerthread-bd54d69fe91a) I found that's a reasonable length if you'd like more information.

Also the delayed execution pattern mentioned in this solution is common on other platforms as well. On iOS, `DispatchQueue.main.async` is similar to a 0-delay post. Even on web, `setTimeout(function () {}, 0)` achieves something similar.

**Comment by indiascope** [comment]
Thank You for the explanation and also for Sharing the Link for Concurrency.

Also good to know, in the newer SDK you may work on fixing this issue internally.  If require any  Help in Testing this issue in Newer SDK Please let me know. It will be my pleasure to help.

**Comment by sshropshire** [comment]
@indiascope thanks for your help! And for sure, we have a PR drafted and will notify here once the fix has been merged-in and released.

**Comment by sshropshire** [comment]
@indiascope np! This is now generally available in version [4.10.0](https://github.com/braintree/braintree_android/blob/master/CHANGELOG.md#4100). Closing this for now. Feel free to re-open if new issues are encountered.
